src/query/functions/temporal/base.go

```diff
@@ -0,0 +1,505 @@
+			if !iter.Next() {
+				return fmt.Errorf("incorrect number of series for block: %d", i)
+			}
+
+			s, err := iter.Current()
+			if err != nil {
+				return err
+			}
+
+			values = append(values, s.Values()...)
+		}
+
+		desiredLength := int(aggDuration / bounds.StepSize)
+		series, err := seriesIter.Current()
+		if err != nil {
+			return err
+		}
+
+		for i := 0; i < series.Len(); i++ {
+			val := series.ValueAtStep(i)
+			values = append(values, val)
+			newVal := math.NaN()
+			if desiredLength <= len(values) {
+				values = values[len(values)-desiredLength:]
+				newVal = c.processor.Process(values)
+			}
+
+			builder.AppendValue(i, newVal)
+
+		}
+	}
+
+	nextBlock := builder.Build()
+	defer nextBlock.Close()
+	return c.controller.Process(nextBlock)
+}
+
+func (c *baseNode) sweep(processedKeys []bool, queryStartBounds block.Bounds, queryEndBounds block.Bounds, maxBlocks int) {
+	prevProcessed := 0
+	numBlocks := len(processedKeys) - 1
+	for i := numBlocks; i >= 0; i-- {
+		processed := processedKeys[i]
+		if !processed {
+			prevProcessed = 0
+			continue
+		}
+
+		dependantBlocks := maxBlocks
+		if dependantBlocks > numBlocks-i-1 {
+			dependantBlocks = numBlocks - i - 1
+		}
+
+		if prevProcessed >= dependantBlocks {
+			c.cache.remove(i)
+		}
+
+		prevProcessed++
+
+	}
+}
+
+// Processor is implemented by the underlying transforms
+type Processor interface {
+	Process(values []float64) float64
+}
+
+// MakeProcessor is a way to create a transform
+type MakeProcessor func(op baseOp, controller *transform.Controller) Processor
+
+type processRequest struct {
+	blk    block.Block
+	bounds block.Bounds
+	deps   []block.Block
+}
+
+// blockCache keeps track of blocks from the same parent across time
+type blockCache struct {
+	mu              sync.Mutex
+	initialized     bool
+	blockList       []block.Block
+	op              baseOp
+	transformOpts   transform.Options
+	startBounds     block.Bounds
+	endBounds       block.Bounds
+	processedBlocks []bool
+}
+
+func newBlockCache(op baseOp, transformOpts transform.Options) *blockCache {
+	return &blockCache{
+		op:            op,
+		transformOpts: transformOpts,
+	}
+}
+func (c *blockCache) init(bounds block.Bounds) {
+	if c.initialized {
+		return
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.initialized {
+		return
+	}
+
+	timeSpec := c.transformOpts.TimeSpec
+	c.startBounds = bounds.Nearest(timeSpec.Start)
+	c.endBounds = bounds.Nearest(timeSpec.End.Add(-1 * bounds.StepSize))
+	numBlocks := c.endBounds.End().Sub(c.startBounds.Start) / bounds.Duration
+	c.blockList = make([]block.Block, numBlocks)
+	c.processedBlocks = make([]bool, numBlocks)
+	c.initialized = true
+}
+
+func (c *blockCache) index(t time.Time) (int, error) {
+	start := c.startBounds.Start
+	if t.Before(start) || t.After(c.endBounds.Start) {
+		return 0, fmt.Errorf("invalid time for the block cache: %v, start: %v, end: %v", t, start, c.endBounds.Start)
+	}
+
+	return int(t.Sub(start) / c.startBounds.Duration), nil
```

Comment:
Add a sanity check for startBounds.Duration==0?

Reply:
will panic in init if that's the case so not needed here. To fix init panic, I added a sanity check in Process()